### ISSUE TITLE
strange kernel call  chains for fib

### ISSUE THREAD

**Issue by jmellorcrummey** [HPCToolkit/hpctoolkit#45]
machine: mic.cs
branch: master
tag: 288e83a4d52e4984d86af12528eaeb3e6e87243d
kernel settings
```
/proc/sys/kernel/perf_event_paranoid=1
/proc/sys/kernel/kptr_restrict=0
```

When kernel sampling is enabled and kptr_restrict=0, we see some strange call chains for fib. This example was dynamically linked.

There seem to be several problems.
1) why are there calls to fib deep in kernel call chains? It doesn't seem that the context is being constructed properly.
2) why are kernel call chains not fused when they seem to come from the same place? Perhaps they are the result of interrupts at different addresses? Should they be fused?

![fib-mic-cycles-dynamic](https://user-images.githubusercontent.com/3926112/34296353-f6a20820-e6d7-11e7-8b12-ad5a33c1533f.png)

Here is a snippet of the call chains from one of the hpcrun files. It shows a PC for fib (load module 1, offset 0x400577)
```
gdb fib-dynamic
(gdb) x/i 0x400577
   0x400577 <fib>:      push   %rbp
```
as a leaf on a call chain with many kernel (load module 20) addresses above it.

```
  [node: (id: 460) (id-parent: 410) (lm-id: 1) (lm-ip: 0x40057c) 
  (metrics: 2.3281e+08)
  ]
  [node: (id: 632) (id-parent: 460) (lm-id: 20) (lm-ip: 0xffffffff9f89cf13) 
  (metrics: 0)
  ]
  [node: (id: 634) (id-parent: 632) (lm-id: 20) (lm-ip: 0xffffffff9f89e548) 
  (metrics: 0)
  ]
  [node: (id: 636) (id-parent: 634) (lm-id: 20) (lm-ip: 0xffffffff9f04e678) 
  (metrics: 0)
  ]
  [node: (id: 638) (id-parent: 636) (lm-id: 20) (lm-ip: 0xffffffff9f11fac0) 
  (metrics: 0)
  ]
  [node: (id: 640) (id-parent: 638) (lm-id: 20) (lm-ip: 0xffffffff9f11f400) 
  (metrics: 0)
  ]
  [node: (id: 642) (id-parent: 640) (lm-id: 20) (lm-ip: 0xffffffff9f12f299) 
  (metrics: 0)
  ]
  [node: (id: 644) (id-parent: 642) (lm-id: 20) (lm-ip: 0xffffffff9f12ec86) 
  (metrics: 0)
  ]
  [node: (id: 646) (id-parent: 644) (lm-id: 20) (lm-ip: 0xffffffff9f11eab7) 
  (metrics: 0)
  ]
  [node: (id: 648) (id-parent: 646) (lm-id: 20) (lm-ip: 0xffffffff9f0d3752) 
  (metrics: 0)
  ]
  [node: (id: 650) (id-parent: 648) (lm-id: 20) (lm-ip: 0xffffffff9f0e0a2f) 
  (metrics: 0)
  ]
  [node: (id: -652) (id-parent: 650) (lm-id: 20) (lm-ip: 0xffffffffffffff80) 
  (metrics: 1.10855e+07)
  ]
  [node: (id: -655) (id-parent: 650) (lm-id: 1) (lm-ip: 0x400577) 
  (metrics: 0)

```

**Comment by jmellorcrummey** [comment]
@laksono I think we should turn off kernel sampling in master until we resolve these issues.
FYI: @unkadoug

**Comment by jmellorcrummey** [comment]
The data collection problem here causes issue #44 . However, #44 should be avoided by realizing that if we can't interpret kernel symbols, we shouldn't collect deep kernel call chains that we can't interpret.

**Comment by laksono** [comment]
This bug is fixed in mult-kernel branch commit be2ba06a959ce57c171546f732034c86eebca837 
see https://github.com/HPCToolkit/hpctoolkit/commit/be2ba06a959ce57c171546f732034c86eebca837

The bug occurs when profiling with a perf_event event and traces.
Reason: we always assume there is no cct modification when calling hpcrun_backtrace2cct() and assume the return leaf_func is a direct descendant of sample node. (the trace will use leaf_func as their frame).
However, this is not true as deep inside hpcrun_backtrace2cct(), we add kernel call chains. Hence the variable leaf_func is not anymore direct descendant of the sample node.

Fix: Using sample node as the trace frame. This works with fib and amg2006 (apparently). Need more tests to validate.
